### PR TITLE
Fix conditions to restart colgen when using lazy cut gen

### PR DESCRIPTION
**Pull request by guimarqu**
Fix: Colgen cannot declare the convergence just after an essential cut has been added to the master.

Now, colgen will restart only if we add an essential cut in phase 2.

We still perform essential cut generation as I said in #452 .

### CODE REVIEW

File: src/Algorithm/colgen.jl [hunk 1]
```diff
@@ -90,18 +90,15 @@ function run!(algo::ColumnGeneration, env::Env, data::ReformData, input::Optimiz
...
 
     restart = true
     while should_do_ph_1(optstate) && restart && !stop
         set_ph1!(master, optstate)
-        stop, restart = cg_main_loop!(algo, env, 1, optstate, data)
-        restart && break
+        stop, _ = cg_main_loop!(algo, env, 1, optstate, data)
         if !stop
             set_ph2!(master, optstate) # pure ph2
             stop, restart = cg_main_loop!(algo, env, 2, optstate, data)
```
Comment (guimarqu) [inline src/Algorithm/colgen.jl:104]: change here

File: src/Algorithm/colgen.jl [hunk 2]
```diff
@@ -633,6 +630,7 @@ function cg_main_loop!(
 
     redcostsvec = ReducedCostsVector(dwspvars, dwspforms)
     iteration = 0
+    essential_cuts_separated = false
```
Comment (guimarqu) [inline src/Algorithm/colgen.jl:633]: new variable here

File: src/Algorithm/colgen.jl [hunk 3]
```diff
@@ -696,8 +694,10 @@ function cg_main_loop!(
                     if cutcb_output.nb_cuts_added == 0
                         update_ip_primal_sol!(cg_optstate, new_primal_sol)
                     else
-                        # because the new cuts may make the master infeasible
-                        return false, true
+                        essential_cuts_separated = true
+                        if phase == 2 # because the new cuts may make the master infeasible
+                            return false, true
+                        end
```
Comment (guimarqu) [inline src/Algorithm/colgen.jl:700]: stop colgen only if we are in phase 2

File: src/Algorithm/colgen.jl [hunk 4]
```diff
@@ -796,6 +796,7 @@ function cg_main_loop!(
             @warn "Maximum number of column generation iteration is reached."
```
Comment (guimarqu) [inline src/Algorithm/colgen.jl:796]: Do not end & declare convergence if an essential cut has been returned just before

**Review by guimarqu — COMMENTED** [review]
Comments to help the reviewer because I formatted the file.

File: src/Algorithm/colgen.jl [hunk 5]
```diff
@@ -267,17 +264,17 @@ function compute_red_cost(
     else
         red_cost = getvalue(spsol)
     end
-    #red_cost -= (spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual)
-    red_cost -= (spinfo.lb_dual + spinfo.ub_dual)
+    # red_cost -= (spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual)
```
Comment (rrsadykov) [inline src/Algorithm/colgen.jl:294]: This commented line can be removed

File: src/Algorithm/colgen.jl [hunk 6]
```diff
@@ -22,15 +22,15 @@ Column generation algorithm. It applies `restr_master_solve_alg` to solve the li
...
-    restr_master_solve_alg = SolveLpForm(get_dual_solution = true)
-    #TODO : pricing problem solver may be different depending on the
+    restr_master_solve_alg = SolveLpForm(get_dual_solution=true)
+    # TODO : pricing problem solver may be different depending on the
     #       pricing subproblem
     pricing_prob_solve_alg = SolveIpForm(
-        deactivate_artificial_vars = false,
-        enforce_integrality = false,
-        log_level = 2
+        deactivate_artificial_vars=false,
```
Comment (rrsadykov) [inline src/Algorithm/colgen.jl:29]: why did you remove spaces around =? This is a new convention? I think with spaces it looks better

File: src/Algorithm/colgen.jl [hunk 7]
```diff
@@ -505,10 +502,10 @@ function update_lagrangian_dual_bound!(
             puremastvars_contrib += redcost * mult
         end
     end
-
+    
     valid_lagr_bound = DualBound{S}(puremastvars_contrib + dualsol.bound)
     for (spuid, spinfo) in spinfos
-        valid_lagr_bound += spinfo.valid_dual_bound_contrib
+    valid_lagr_bound += spinfo.valid_dual_bound_contrib
```
Comment (rrsadykov) [inline src/Algorithm/colgen.jl:494]: Indentation is broken

File: src/Algorithm/colgen.jl [hunk 8]
```diff
@@ -517,7 +514,7 @@ function update_lagrangian_dual_bound!(
     if stabilization_is_used(algo)
         pseudo_lagr_bound = DualBound{S}(puremastvars_contrib + dualsol.bound)
         for (spuid, spinfo) in spinfos
-            pseudo_lagr_bound += spinfo.pseudo_dual_bound_contrib
+pseudo_lagr_bound += spinfo.pseudo_dual_bound_contrib
```
Comment (rrsadykov) [inline src/Algorithm/colgen.jl:503]: Indentation is broken

File: src/Algorithm/colgen.jl (hunk 3 above)
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:677]: As I mentioned in other pull request, 'dwsprep_coefmatrix' should be recalculated, if essential cuts are added and column generation is not stopped

**Review by rrsadykov — CHANGES_REQUESTED** [review]

File: src/Algorithm/colgen.jl (hunk 6 above)
Comment (guimarqu) [reply src/Algorithm/colgen.jl:29]: It's the formatter and it looks like to be the convention : https://github.com/invenia/BlueStyle#keyword-arguments

**Review by rrsadykov — APPROVED** [review]
